File: frontend-app/src/app/services/employee.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Employee {
  id?: number;
  firstName: string;
  lastName: string;
  address?: string;
  salary: number;
}

export interface EmployeeSalaryStatistics {
  averageSalary: number;
  medianSalary: number;
  minSalary: number;
  maxSalary: number;
  totalEmployees: number;
}

export interface SalaryRangeCount {
  rangeName: string;
  minSalary: number;
  maxSalary: number;
  count: number;
}

export interface EmployeeStatistics {
  totalEmployees: number;
  averageSalary: number;
  distinctCities: number;
  highEarners: number;
}

@Injectable({
  providedIn: 'root'
})
export class EmployeeService {
  private baseUrl = 'http://localhost:8081/api/v1/employees';

  constructor(private http: HttpClient) {}

  getAllEmployees(): Observable<Employee[]> {
    return this.http.get<Employee[]>(this.baseUrl);
  }

  getEmployeeById(id: number): Observable<Employee> {
    return this.http.get<Employee>(`${this.baseUrl}/${id}`);
  }

  createEmployee(employee: Employee): Observable<Employee> {
    return this.http.post<Employee>(this.baseUrl, employee);
  }

  updateEmployee(id: number, employee: Employee): Observable<Employee> {
    return this.http.put<Employee>(`${this.baseUrl}/${id}`, employee);
  }

  deleteEmployee(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }

  searchEmployeesByLastName(lastName: string): Observable<Employee[]> {
    const params = new HttpParams().set('lastName', lastName);
    return this.http.get<Employee[]>(`${this.baseUrl}/search/lastname`, { params });
  }

  searchEmployeesByFirstName(firstName: string): Observable<Employee[]> {
    const params = new HttpParams().set('firstName', firstName);
    return this.http.get<Employee[]>(`${this.baseUrl}/search/firstname`, { params });
  }

  searchEmployeesByFullName(fullName: string): Observable<Employee[]> {
    const params = new HttpParams().set('fullName', fullName);
    return this.http.get<Employee[]>(`${this.baseUrl}/search/fullname`, { params });
  }

  getEmployeesBySalaryRange(minSalary: number, maxSalary: number): Observable<Employee[]> {
    const params = new HttpParams()
      .set('minSalary', minSalary.toString())
      .set('maxSalary', maxSalary.toString());
    return this.http.get<Employee[]>(`${this.baseUrl}/filter/salary-range`, { params });
  }

  getHighEarners(minSalary: number): Observable<Employee[]> {
    const params = new HttpParams().set('minSalary', minSalary.toString());
    return this.http.get<Employee[]>(`${this.baseUrl}/filter/high-earners`, { params });
  }

  searchEmployeesByAddress(address: string): Observable<Employee[]> {
    const params = new HttpParams().set('address', address);
    return this.http.get<Employee[]>(`${this.baseUrl}/search/address`, { params });
  }

  getEmployeesOrderedBySalary(): Observable<Employee[]> {
    return this.http.get<Employee[]>(`${this.baseUrl}/ordered/salary`);
  }

  getEmployeesOrderedByName(): Observable<Employee[]> {
    return this.http.get<Employee[]>(`${this.baseUrl}/ordered/name`);
  }

  searchEmployeesAdvanced(firstName?: string, lastName?: string, minSalary?: number, maxSalary?: number): Observable<Employee[]> {
    let params = new HttpParams();
    if (firstName) params = params.set('firstName', firstName);
    if (lastName) params = params.set('lastName', lastName);
    if (minSalary) params = params.set('minSalary', minSalary.toString());
    if (maxSalary) params = params.set('maxSalary', maxSalary.toString());
    
    return this.http.get<Employee[]>(`${this.baseUrl}/search/advanced`, { params });
  }

  getSalaryStatistics(): Observable<EmployeeSalaryStatistics> {
    return this.http.get<EmployeeSalaryStatistics>(`${this.baseUrl}/statistics/salary`);
  }

  getTopEarners(limit: number = 10): Observable<Employee[]> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<Employee[]>(`${this.baseUrl}/top-earners`, { params });
  }

  getEmployeeCountBySalaryRanges(): Observable<SalaryRangeCount[]> {
    return this.http.get<SalaryRangeCount[]>(`${this.baseUrl}/statistics/salary-ranges`);
  }

  getEmployeeStatistics(): Observable<EmployeeStatistics> {
    return this.http.get<EmployeeStatistics>(`${this.baseUrl}/statistics`);
  }
}

File: backend-api/src/main/java/com/icefaces/modernization/api/service/EmployeeService.java
package com.icefaces.modernization.api.service;

import com.icefaces.modernization.api.dto.EmployeeDto;
import com.icefaces.modernization.api.dto.EmployeeSalaryStatisticsDto;
import com.icefaces.modernization.api.dto.EmployeeStatisticsDto;
import com.icefaces.modernization.api.dto.SalaryRangeCountDto;
import java.util.List;
import java.util.Optional;

/**
 * Service interface for Employee operations
 * Replaces business logic from ICEfaces managed beans like:
 * - org.icefaces.samples.showcase.example.compat.series.SeriesBean
 * - HR-related form and data management beans
 * - Employee chart and reporting beans
 * 
 * This interface defines the contract for employee-related operations,
 * focusing on HR management and reporting functionality.
 */
public interface EmployeeService {
    
    /**
     * Retrieve all employees
     * Replaces: ICEfaces managed bean employee list initialization
     */
    List<EmployeeDto> getAllEmployees();
    
    /**
     * Find employee by ID
     * Replaces: ICEfaces managed bean selection logic
     */
    Optional<EmployeeDto> getEmployeeById(Long id);
    
    /**
     * Create new employee
     * Replaces: ICEfaces managed bean add employee functionality
     */
    EmployeeDto createEmployee(EmployeeDto employeeDto);
    
    /**
     * Update existing employee
     * Replaces: ICEfaces managed bean update employee functionality
     */
    EmployeeDto updateEmployee(Long id, EmployeeDto employeeDto);
    
    /**
     * Delete employee
     * Replaces: ICEfaces managed bean delete employee functionality
     */
    void deleteEmployee(Long id);
    
    /**
     * Search employees by last name
     * Replaces: ICEfaces employee search functionality
     */
    List<EmployeeDto> searchEmployeesByLastName(String lastName);
    
    /**
     * Search employees by first name
     * Replaces: ICEfaces employee search functionality
     */
    List<EmployeeDto> searchEmployeesByFirstName(String firstName);
    
    /**
     * Search employees by full name (first + last)
     * Replaces: ICEfaces comprehensive name search
     */
    List<EmployeeDto> searchEmployeesByFullName(String fullName);
    
    /**
     * Find employees within salary range
     * Replaces: ICEfaces salary filter components
     */
    List<EmployeeDto> getEmployeesBySalaryRange(Integer minSalary, Integer maxSalary);
    
    /**
     * Find high earners (above specified salary)
     * Replaces: ICEfaces business logic for high earner reports
     */
    List<EmployeeDto> getHighEarners(Integer minSalary);
    
    /**
     * Search employees by address
     * Replaces: ICEfaces location-based employee search
     */
    List<EmployeeDto> searchEmployeesByAddress(String addressText);
    
    /**
     * Get employees ordered by salary (descending)
     * Replaces: ICEfaces salary ranking functionality
     */
    List<EmployeeDto> getEmployeesOrderedBySalary();
    
    /**
     * Get employees ordered alphabetically by last name
     * Replaces: ICEfaces alphabetical listing
     */
    List<EmployeeDto> getEmployeesOrderedByName();
    
    /**
     * Advanced employee search with multiple criteria
     * Replaces: Complex ICEfaces filter combinations
     */
    List<EmployeeDto> searchEmployees(String firstName, String lastName, 
                                    Integer minSalary, Integer maxSalary);
    
    /**
     * Get salary statistics for reporting and dashboards
     * Replaces: ICEfaces statistical calculation logic
     */
    EmployeeSalaryStatisticsDto getSalaryStatistics();
    
    /**
     * Get top earners (limited list for executive reports)
     * Replaces: ICEfaces business logic for top performer reports
     */
    List<EmployeeDto> getTopEarners(int limit);
    
    /**
     * Get employee count by salary ranges (for chart generation)
     * Replaces: ICEfaces chart data preparation logic
     */
    List<SalaryRangeCountDto> getEmployeeCountBySalaryRanges();
    
    /**
     * Calculate employee statistics for HR reporting
     * Replaces: ICEfaces HR analytics functionality
     */
    EmployeeStatisticsDto getEmployeeStatistics();
}